sirepo/auth_db/__init__.py

```diff
@@ -0,0 +1,294 @@
+# -*- coding: utf-8 -*-
+"""Auth database
+
+:copyright: Copyright (c) 2018-2019 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdc, pkdexc, pkdlog, pkdp
+import contextlib
+import importlib
+import inspect
+import pykern.pkconfig
+import pykern.pkio
+import pykern.pkinspect
+import sirepo.auth_role
+import sirepo.feature_config
+import sirepo.quest
+import sirepo.srdb
+import sirepo.srtime
+import sirepo.util
+import sqlalchemy
+import sqlalchemy.ext.declarative
+import sqlalchemy.orm
+import sqlalchemy.sql.expression
+import subprocess
+
+
+#: sqlite file located in sirepo_db_dir
+_SQLITE3_BASENAME = "auth.db"
+
+#: SQLAlchemy database engine
+_engine = None
+
+STRING_ID = sqlalchemy.String(8)
+STRING_NAME = sqlalchemy.String(100)
+
+_models = None
+
+
+@sqlalchemy.ext.declarative.as_declarative()
+class UserDbBase:
+    def __init__(self, **kwargs):
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+
+    @classmethod
+    def add_column_if_not_exists(cls, table, column, column_type):
+        column_type = column_type.upper()
+        t = table.__table__.name
+        r = cls._execute_raw_sql(f"PRAGMA table_info({t})")
+        for c in r.all():
+            if not c[1] == column:
+                continue
+            assert c[2] == column_type, (
+                f"unexpected column={c} when adding column={column} of",
+                f" type={column_type} to table={table}",
+            )
+            return
+        cls._execute_raw_sql(f"ALTER TABLE {t} ADD {column} {column_type}")
```

Comment:
I think a comment here that this is never used with user defined data could be good and maybe moving _execute_raw_sql into this function so it is never used elsewhere.

Reply:
Thanks. I read about this, and we should just use placeholders, which seem to work fine for DDL. 57dd3717a